packages/proto-rpc/src/ClientTransport.ts

```diff
@@ -0,0 +1,132 @@
+import {
+    ClientStreamingCall,
+    ServerStreamingCall,
+    DuplexStreamingCall,
+    Deferred,
+    RpcTransport,
+    MethodInfo,
+    RpcError,
+    RpcOptions,
+    RpcMetadata,
+    RpcStatus,
+    UnaryCall,
+    mergeRpcOptions
+} from '@protobuf-ts/runtime-rpc'
+import { v4 } from 'uuid'
+import { RpcMessage } from './proto/ProtoRpc'
+import EventEmitter = require('events')
+import { Logger } from './Logger'
+
+export enum Event {
+    RPC_REQUEST = 'rpcRequest'
+}
+
+export interface ClientTransport {
+    on(event: Event.RPC_REQUEST, listener: (results: ResultParts, rpcMessage: RpcMessage, options: ProtoRpcOptions) => void): this
+}
+
+export interface ResultParts {
+    header: Deferred<RpcMetadata>,
+    message: Deferred<object>,
+    status: Deferred<RpcStatus>,
+    trailer: Deferred<RpcMetadata>,
+    messageParser: (bytes: Uint8Array) => object
+}
+
+export interface ProtoRpcOptions extends RpcOptions {
+    notification?: boolean
+}
+
+const logger = new Logger(module)
+
+export class ClientTransport extends EventEmitter implements RpcTransport {
+    private static objectCount = 0
+    private readonly objectId: number
+    protected readonly defaultOptions: ProtoRpcOptions
+
+    constructor(defaultTimeout = 5000) {
+        super()
+        this.objectId = ClientTransport.objectCount++
+        this.defaultOptions = {
+            timeout: defaultTimeout,
+            clientId: this.objectId
+        }
+    }
+
+    mergeOptions(options?: Partial<ProtoRpcOptions>): ProtoRpcOptions {
+        return mergeRpcOptions(this.defaultOptions, options)
+    }
+
+    private createRequestHeaders(method: MethodInfo, notification?: boolean): {
+        method: string,
+        request: string,
```

Comment:
Would it make sense to annotate this object with strict types? E.g. what values the request field can get (`request | ...`), and what values `notification` can get. 

There is maybe a reason why `notification` is not a boolean?

Reply:
All of the custom header fields are of type {string, string} currently